[GCS] Optimize `GetAllJobInfo` API for performance

liuxsh9 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

The `GetAllJobInfo` API is frequently called by job, log management, and dashboard data collection interfaces. Currently, the API retrieves three fields: 1) job table data, 2) is_running_tasks, and 3) JobInfo. The latter two fields are costly, but they appear to be unused (based on my knowledge and search). To improve performance, we added two optional parameters to control whether to retrieve those two fields (defaulting to not). This optimization reduces the API response time from 1s to approximately 0.05s when there are hundreds of jobs in the cluster. If the latter two fields are indeed unnecessary, consider removing the collection operations altogether.

![image](https://github.com/user-attachments/assets/1e9b5b88-0991-424c-97ff-8b6f4c4bcc93)


## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rynewang commented on python/ray/_raylet.pyx:
```diff
@@ -2907,18 +2907,23 @@ cdef class OldGcsClient:
 
     @_auto_reconnect
-    def get_all_job_info(self, timeout=None) -> Dict[JobID, JobTableData]:
+    def get_all_job_info(self, timeout=None, **kwargs) -> Dict[JobID, JobTableData]:
```
wonder if we can make the fields just defined here, instead of **kwargs:

def get_all_job_info(self, query_job_info_field, c_query_is_running_tasks_field, timeout=None) -> Dict[JobID, JobTableData]:

alexeykudinkin replied:
+1 

Let's err on the side of properly codifying params

liuxsh9 replied:
Good suggestion, have updated to use explicit field definitions instead.

rynewang commented on python/ray/includes/gcs_client.pxi:
```diff
@@ -433,20 +433,28 @@ cdef class NewGcsClient:
 
     def get_all_job_info(
-        self, timeout: Optional[float] = None
+        self, timeout: Optional[float] = None, **kwargs
```
ditto, explicit args

liuxsh9 replied:
Done.

rynewang commented on src/ray/gcs/gcs_server/gcs_job_manager.cc:
```diff
@@ -203,79 +203,91 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
-        reply->mutable_job_info_list(i)->set_is_running_tasks(false);
-        core_worker_clients_.Disconnect(worker_id);
+      if (request.has_query_is_running_tasks_field() &&
+          request.query_is_running_tasks_field()) {
```
you can omit the has_ check. if it's not set, query_is_running_tasks_field() defaults to false.

liuxsh9 replied:
Good suggestion!

rynewang reviewed (commented):
Generally LGTM! One issue though: the field is used by places like:

[get_total_num_running_jobs_to_report](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/_private/usage/usage_lib.py#L540)

[@routes.get("/api/jobs/")](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/job/job_head.py#L460)

[/api/jobs/{job_or_submission_id}](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/job/job_head.py#L439)

[get_job_info](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/util/state/state_manager.py#L401)

And those APIs may be consumed by external code. To keep backwards compatibility can you add args to the web APIs with default fetch=True? Then in call sites like https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/log/log_manager.py#L150 we can set fetch=False.

Especially, we have a [`@PublicAPI` ](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/job/sdk.py#L324), @jjyao can you help evaluate if this PublicAPI contained such fields, and shall we change its semantics for this better perf?

alexeykudinkin commented on python/ray/_private/gcs_aio_client.py:
```diff
@@ -86,7 +87,8 @@ def __init__(self, inner, loop, executor):
     def _function_to_async(self, func):
         async def wrapper(*args, **kwargs):
-            return await self.loop.run_in_executor(self.executor, func, *args, **kwargs)
+            partial_func = partial(func, *args, **kwargs)
```
+1

Let's avoid bundling up unrelated changes

liuxsh9 replied:
We discovered that `run_in_executor` doesn't support `**kwargs` [ref: asyncio.loop.run_in_executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor). The implicit params passing was exposing this issue. For now, we've changed to explicit params passing. So have reverted this change. 
Anyway, should we adjust the input params to accommodate `run_in_executor`'s limitations?

liuxsh9 commented:
> Generally LGTM! One issue though: the field is used by places like:
> 
> [get_total_num_running_jobs_to_report](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/_private/usage/usage_lib.py#L540)
> 
> [@routes.get("/api/jobs/")](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/job/job_head.py#L460)
> 
> [/api/jobs/{job_or_submission_id}](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/job/job_head.py#L439)
> 
> [get_job_info](https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/util/state/state_manager.py#L401)
> 
> And those APIs may be consumed by external code. To keep backwards compatibility can you add args to the web APIs with default fetch=True? Then in call sites like
> 
> https://github.com/ray-project/ray/blob/c083ea53b9c1ac9639a864fa3cbc59bd4a56181a/python/ray/dashboard/modules/log/log_manager.py#L150
> 
> we can set fetch=False.

Sorry, I am a little confused. Do you mean we're still fetching all 3 fields by default, and then setting `query_x_field=False` in individual APIs that definitely don't need them?

jjyao commented:
> Sorry, I am a little confused. Do you mean we're still fetching all 3 fields by default, and then setting query_x_field=False in individual APIs that definitely don't need them?

Yes.

liuxsh9 commented:
Updated the default to retrieving all fields, like before. To accommodate proto3's boolean fields defaulting to false, we've changed the params to `skip_x_field`. So now we can set `skip_x_field=True` when method doesn't need the field. Additionally, we've optimized functions like `get_driver_jobs` to use `GetAllJobInfo` with selective field retrieval.

jjyao commented on src/ray/gcs/gcs_client/accessor.cc:
```diff
@@ -82,7 +82,10 @@ Status JobInfoAccessor::AsyncSubscribeAll(
     };
-    RAY_CHECK_OK(AsyncGetAll(callback, /*timeout_ms=*/-1));
+    RAY_CHECK_OK(AsyncGetAll(callback,
+                             /*skip_job_info_field=*/false,
+                             /*skip_is_running_tasks_field=*/false,
```
Can those be True as well If they are not used?

liuxsh9 replied:
Yes!

jjyao commented on src/ray/gcs/gcs_client/global_state_accessor.cc:
```diff
@@ -68,6 +68,8 @@ std::vector<std::string> GlobalStateAccessor::GetAllJobInfo() {
     absl::ReaderMutexLock lock(&mutex_);
     RAY_CHECK_OK(gcs_client_->Jobs().AsyncGetAll(
         TransformForMultiItemCallback<rpc::JobTableData>(job_table_data, promise),
+        /*skip_job_info_field=*/false,
+        /*skip_is_running_tasks_field=*/false,
```
Can those be True? Are those two fields used?

liuxsh9 replied:
`GlobalStateAccessor::GetAllJobInfo()` is invoked by multiple functions, including

https://github.com/ray-project/ray/blob/f40313b0acc50a60957d1ecd2b75946ca6079ce4/python/ray/_private/state.py#L170

which explicitly sets two params to True as it doesn't utilize those fields.

And

https://github.com/ray-project/ray/blob/f40313b0acc50a60957d1ecd2b75946ca6079ce4/java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java#L50

may be called by external methods, so we're leaving it unchanged and defaulting to False.

To accommodate these varying callers, have added explicit `skip` params to `GetAllJobInfo()`.

jjyao reviewed (commented):
lg, thanks!

jjyao commented:
Failed tests are probably related.

jjyao commented on python/ray/_private/gcs_aio_client.py:
```diff
@@ -205,9 +205,16 @@ async def internal_kv_keys(
     async def get_all_job_info(
         self,
         job_or_submission_id: Optional[str] = None,
```
nit: since there are more parameters, let's force them to be keyword arguments by doing

```
get_all_job_info(self, *, job_or_submission_id, ...)
```

jjyao commented on python/ray/_private/state.py:
```diff
@@ -181,7 +181,7 @@ def job_table(self):
         self._check_connected()
 
-        job_table = self.global_state_accessor.get_job_table()
+        job_table = self.global_state_accessor.get_job_table(True, True)
```
Using keyword arguments otherwise it's hard to know what these two True are

jjyao commented on python/ray/_raylet.pyx:
```diff
@@ -2907,8 +2907,10 @@ cdef class OldGcsClient:
-    def get_all_job_info(self, job_or_submission_id: str = None,
-                         timeout=None) -> Dict[JobID, JobTableData]:
+    def get_all_job_info(
+        self, job_or_submission_id: str = None, skip_submission_job_info_field=False,
```
```suggestion
        self, *, job_or_submission_id: str = None, skip_submission_job_info_field=False,
```

jjyao commented on python/ray/includes/global_state_accessor.pxi:
```diff
@@ -48,10 +48,16 @@ cdef class GlobalStateAccessor:
 
-    def get_job_table(self):
+    def get_job_table(
+        self, skip_submission_job_info_field=False, skip_is_running_tasks_field=False
```
```suggestion
        self, *, skip_submission_job_info_field=False, skip_is_running_tasks_field=False
```

jjyao commented on src/ray/gcs/gcs_server/gcs_job_manager.cc:
```diff
@@ -288,78 +287,90 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
+          reply->mutable_job_info_list(i)->set_is_running_tasks(false);
+          core_worker_clients_.Disconnect(worker_id);
+          (*num_processed_jobs)++;
+          ;
```
This can be deleted

jjyao reviewed (commented):
Looks great!

The remaining comments are about changing them to keyword arguments.

jjyao commented on python/ray/includes/gcs_client.pxi:
```diff
@@ -434,11 +434,15 @@ cdef class NewGcsClient:
 
     def get_all_job_info(
         self, job_or_submission_id: Optional[str] = None,
```
```suggestion
        self, *, job_or_submission_id: Optional[str] = None,
```

jjyao commented on python/ray/includes/gcs_client.pxi:
```diff
@@ -447,17 +451,22 @@ cdef class NewGcsClient:
         return raise_or_return((convert_get_all_job_info(status, move(reply))))
 
     def async_get_all_job_info(
         self, job_or_submission_id: Optional[str] = None,
```
```suggestion
        self, *, job_or_submission_id: Optional[str] = None,
```

liuxsh9 commented:
Updates applied, kindly review at your convenience. @jjyao
